test/mpi/rma/lock_contention_dt.c

```diff
@@ -8,90 +8,138 @@
 
+            for (j = 0; j < send_dtp->DTP_num_objs; j++) {
+                err = DTP_obj_create(send_dtp, j, 0, 1, count);
+                if (err != DTP_SUCCESS) {
+                    MPI_Type_get_name(basic_type, type_name, &len);
+                    fprintf(stdout, "Error while creating obj %d in send pool (%s,%d)\n", j,
+                            type_name, count);
+                    fflush(stdout);
+                    break;
+                }
+
+                sendcount = send_dtp->DTP_obj_array[j].DTP_obj_count;
+                sendtype = send_dtp->DTP_obj_array[j].DTP_obj_type;
+                sendbuf = send_dtp->DTP_obj_array[j].DTP_obj_buf;
+
+                if (rank != target) {
                     MPI_Win_lock(MPI_LOCK_EXCLUSIVE, target, 0, win);
-                    MPI_Accumulate(sendtype.buf, sendtype.count,
-                                   sendtype.datatype, target, 0,
-                                   recvtype.count, recvtype.datatype, MPI_REPLACE, win);
+                    MPI_Accumulate(sendbuf, sendcount,
+                                   sendtype, target, 0, recvcount, recvtype, MPI_REPLACE, win);
                     MPI_Win_unlock(target, win);
                     MPI_Barrier(comm);
 
-                    char *resbuf = (char *) calloc(lb + extent * recvtype.count, sizeof(char));
+                    char *resbuf = (char *) calloc(lb + extent * recvcount, sizeof(char));
 
                     /*wait for the destination to finish checking and reinitializing the buffer */
                     MPI_Barrier(comm);
 
                     MPI_Win_lock(MPI_LOCK_EXCLUSIVE, target, 0, win);
-                    MPI_Get_accumulate(sendtype.buf, sendtype.count,
-                                       sendtype.datatype, resbuf, recvtype.count, recvtype.datatype,
-                                       target, 0, recvtype.count, recvtype.datatype, MPI_REPLACE,
-                                       win);
+                    MPI_Get_accumulate(sendbuf, sendcount,
+                                       sendtype, resbuf, recvcount, recvtype,
+                                       target, 0, recvcount, recvtype, MPI_REPLACE, win);
                     MPI_Win_unlock(target, win);
                     MPI_Barrier(comm);
                     free(resbuf);
-                } else {        /* Target checks the result */
-                    int err;
+                }
+                else {  /* Target checks the result */
                     MPI_Barrier(comm);
                     MPI_Win_lock(MPI_LOCK_EXCLUSIVE, rank, 0, win);
-                    err = MTestCheckRecv(0, &recvtype);
-                    if (err)
+                    err = DTP_obj_buf_check(recv_dtp, i, 0, 1, count);
+                    if (err != DTP_SUCCESS) {
                         errs++;
-                    recvtype.InitBuf(&recvtype);
+                    }
+                    memset(recvbuf, 0, lb + extent * recvcount);
```

Comment:
The buffer is firstly reset by `DTP_obj_create`, but then by `memset` in the test code. It looks inconsistent to me. Maybe you can add a `set_buf(start, stride, count)` interface ?

Reply:
I will remove memsets